File: felix.js
(function (ext) {
	var socket = null;
	var connected = false;
	var digital_inputs = new Array(32); // an array to hold possible digital input values for the reporter block
	var myStatus = 1; // initially yellow
	var myMsg = 'not_ready';
	var motor_configured = false;
	var ir_sensor_tripped = false;
    
	ext.connect = function (callback) {
		console.log("Starting websocket to port 9001");
		window.socket = new WebSocket("ws://127.0.0.1:9001");
		window.socket.onopen = function () {
			var msg = JSON.stringify({
				"command": "ready"
			});
			window.socket.send(msg);
			myStatus = 2;
			// change status light from yellow to green
			myMsg = 'ready';
			connected = true;
			// initialize the reporter buffer
			digital_inputs.fill('0');
			// give the connection time establish
			window.setTimeout(function() {
				callback();
			}, 1000);
		};
        
		window.socket.onmessage = function (message) {
			var msg = JSON.parse(message.data);
			// handle the only reporter message from the server
			// for changes in digital input state
			var reporter = msg['report'];
			if(reporter === 'digital_input_change') {
				var pin = msg['pin'];
				digital_inputs[parseInt(pin)] = msg['level'];
			}
			if(reporter === 'ir_sensor_triggered') {
				var trip = parseInt(msg['state']);
				ir_sensor_tripped = (trip == 1);
			}
			console.log(message.data);
		};

		window.socket.onclose = function (e) {
			console.log("Connection closed.");
			socket = null;
			connected = false;
			myStatus = 1;
			myMsg = 'not_ready';
			motor_configured = false;
		};
	};
	
	ext.disconnect = function (callback) {
		console.log("Disconnecting websocket on port 9001");
		myMsg = "disconnected";
		connected = false;

		var msg = JSON.stringify({
			"command": "shutdown"
		});
		window.socket.send(msg);
	};

	// Cleanup function when the extension is unloaded
	ext._shutdown = function () {
		var msg = JSON.stringify({
			"command": "shutdown"
		});
		window.socket.send(msg);
	};
    
	// Status reporting code
	// Use this to report missing hardware, plugin or unsupported browser
	ext._getStatus = function (status, msg) {
		return {status: myStatus, msg: myMsg};
	};
    
	ext.input = function(pin) {
		if (connected == false) {
			alert("Server Not Connected");
		}
		// validate the pin number for the mode
		if (validatePin(pin)) {
			var msg = JSON.stringify({
				"command": 'input', 'pin': pin
			});
			sendMessage(msg);
		}
	};
    
	// when the digital write block is executed
	ext.digital_write = function (pin, state) {
		if (connected == false) {
			alert("Server Not Connected");
		}
		console.log("digital write");
		// validate the pin number for the mode
		if (validatePin(pin)) {
			var msg = JSON.stringify({
				"command": 'digital_write', 'pin': pin, 'state': state
			});
			sendMessage(msg);
		}
	};
    
	// when the PWM block is executed
	ext.analog_write = function (pin, value) {
		if (connected == false) {
			alert("Server Not Connected");
		}
		console.log("analog write");
		// validate the pin number for the mode
		if (validatePin(pin)) {
			// validate value to be between 0 and 255
			if (value === 'VAL') {
				alert("PWM Value must be in the range of 0 - 255");
			}
			else {
				value = parseInt(value);
				if (value < 0 || value > 255) {
					alert("PWM Value must be in the range of 0 - 255");
				}
				else {
					var msg = JSON.stringify({
						"command": 'analog_write', 'pin': pin, 'value': value
					});
					sendMessage(msg);
				}
			}
		}
	};

	// ***Hackeduca --> when the Servo block is executed
	ext.servo = function (pin, value) {
		if (connected == false) {
			alert("Server Not Connected");
		}
		console.log("servo");
		// validate the pin number for the mode
		if (validatePin(pin)) {
			// validate value to be between 0° and 180°
			if (value === 'VAL') {
				alert("Servo Value must be in the range of 0° - 180°");
			}
			else {
				value = parseInt(value);
				if (value < 0 || value > 180) {
					alert("Servo Value must be in the range of 0° - 180°");
				}
				else {
					var msg = JSON.stringify({
						"command": 'servo', 'pin': pin, 'value': value
					});
					sendMessage(msg);
				}
			}
		}
	};

	ext.play_tone = function (pin, frequency) {
		if (connected == false) {
			alert("Server Not Connected");
		}
		// validate the pin number for the mode
		if (validatePin(pin)) {
			var msg = JSON.stringify({
				"command": 'tone', 'pin': pin, 'frequency': frequency
			});
			sendMessage(msg);
		}
	};

	ext.set_lantern_brightness = function(brightness) {
		if (connected == false) {
			alert("Server Not Connected");
		}
		
		var msg = JSON.stringify({
			"command": 'lantern_brightness', 'brightness': brightness
		});
		sendMessage(msg);
	};

	ext.digital_read = function(pin) {
		console.log("digital_read", pin);
		console.log(digital_inputs);
		return digital_inputs[parseInt(pin)];
	};

	ext.set_led_brightness = function(pin, brightness) {
		if (connected == false) {
			alert("Server Not Connected");
		}

		var msg = JSON.stringify({
			"command": 'led_brightness', 'pin': pin, 'brightness': brightness
		});
		sendMessage(msg);
	};

	ext.set_lantern_to_cycle = function() {
		if (connected == false) {
			alert("Server Not Connected");
		}
		var msg = JSON.stringify({
			"command": 'lantern_cycle_begin'
		});
		sendMessage(msg);
	};

	ext.pause_lantern_cycle = function() {
		if (connected == false) {
			alert("Server Not Connected");
		}
		var msg = JSON.stringify({
			"command": 'lantern_cycle_pause'
		});
		sendMessage(msg);
	};

	ext.set_train_switch = function(pin, direction) {
		if (connected == false) {
			alert("Server Not Connected");
		}
		var msg = JSON.stringify({
			"command": 'set_train_switch', "pin": pin, "direction": direction
		});
		sendMessage(msg);
	};

	ext.console_log = function(message) {
		if (connected == false) {
			alert("Server Not Connected");
		}
		var msg = JSON.stringify({
			"command": 'console_log', "message": message
		});
		sendMessage(msg);
	};

	ext.ir_sensor_activated = function() {
		if (ir_sensor_tripped) {
			console.log("********** Sensor tripped");
			return true;
		} else {
			console.log("++++++++++ Sensor deactivated");
			return false;
		}
	};

	ext.setup_ir_sensor = function(pin) {
		if (connected == false) {
			alert("Server Not Connected");
		}
		var msg = JSON.stringify({"command": 'configure_ir_sensor', "pin": pin});
		sendMessage(msg);
	};

	function validatePin(pin) {
		var rValue = true;
		if (pin === 'PIN') {
			alert("Insert a valid BCM pin number.");
			rValue = false;
		}
		else {
			var pinInt = parseInt(pin);
			if (pinInt < 0 || pinInt > 31) {
				alert("BCM pin number must be in the range of 0-31.");
				rValue = false;
			}
		}
		return rValue;
	}
	
	function validateConnection() {
		if (connected == false) {
			// alert("Server Not Connected");
		}
	}
	
	function sendMessage(msg) {
		console.log(msg);
		window.socket.send(msg);
	}
    
	ext.rotate_motor = function (dir, speed, steps) {
		validateConnection();
		console.log("rotate_motor");
		// alert("SPEED: '" + speed + "', DIR: '" + dir + "', STEPS: '" + steps + "'.");
		
		if (!motor_configured) {
			// alert("Configure motor with four pins before using it.");
			return;
		}
		
		if (steps < 1) {
			// alert("Steps must be > 0");
			return;
		}
		
		if (! (dir === 'cw' || dir === 'ccw')) {
			// alert("Direction must be 'cw' or 'ccw' (It is '" + dir + "'.)");
			return;
		}
		
		if (! (speed === 'slow' || speed === 'medium' || speed === 'fast' || speed === 'super fast')) {
			// alert("Speed must be 'slow', 'medium', or 'fast' (It is '" + speed + "'.)" );
			return;
		}
        
		var msg = JSON.stringify({
			"command": 'rotate_motor', 'speed': speed, 'dir': dir, 'steps': steps
		});
		sendMessage(msg);
	};
	
	ext.setup_motor = function (pin1, pin2, pin3, pin4) {
		var pinsValid = true;
		var pin1Int = parseInt(pin1);
		if (!validatePin(pin1Int)) {
			pinsValid = false;
		}
		var pin2Int = parseInt(pin2);
		if (!validatePin(pin2Int)) {
			pinsValid = false;
		}
		var pin3Int = parseInt(pin3);
		if (!validatePin(pin3Int)) {
			pinsValid = false;
		}
		var pin4Int = parseInt(pin4);
		if (!validatePin(pin4Int)) {
			pinsValid = false;
		}
		
		if (!pinsValid) { return; }
		
		var msg = JSON.stringify({
			"command": 'setup_motor', 'pin1': parseInt(pin1), 'pin2': parseInt(pin2), 'pin3': parseInt(pin3), 'pin4': parseInt(pin4)
		});
		sendMessage(msg);
		motor_configured = true;
	};
    
	// Block and block menu descriptions
	var descriptor = {
		blocks: [
			["w", 'Connect to felix server.', 'connect'],
			["w", 'Disconnect from felix server.', 'disconnect'],
			[" ", 'Set up stepper motor with GPIO pins %n, %n, %n, and %n', "setup_motor", "21", "20", "16", "12"],
			[" ", 'Set magic lantern brightness to %n (0-255)', "set_lantern_brightness", "100"],
			[" ", 'Cycle magic lantern', "set_lantern_to_cycle"],
			[" ", 'Pause magic lantern cycle', "pause_lantern_to_cycle"],
			[" ", 'Set LED on pin %n brightness to %n (0-255)', "set_led_brightness", "PIN", "100"],
			//                 "command": 'setup_motor', 'pin1': pin, 'pin2': pin, 'pin3': pin, 'pin4': pin,
			[" ", 'Rotate stepper motor %m.motor_direction at %m.motor_speed speed for %n steps', "rotate_motor", "DIR", "SPEED", 0],
			[" ", 'Set train switch on pin %n to %m.train_switch_direction', "set_train_switch", 23, "straight"],
			// 				   "command": 'rotate_motor', 'speed': speed, 'dir': dir
			//
			[" ", 'Set pin %n as an input', 'input','PIN'],
			[" ", 'Set up pin %n as an IR sensor', 'setup_ir_sensor','PIN'],
			//             [" ", "Set BCM %n Output to %m.high_low", "digital_write", "PIN", "0"],
			//             [" ", "Set BCM PWM Out %n to %n", "analog_write", "PIN", "VAL"],
			// [" ", "Set BCM %n as Servo with angle = %n (0° - 180°)", "servo", "PIN", "0"],     // ***Hackeduca --> Block for Servo
			//             [" ", "Tone: BCM %n HZ: %n", "play_tone", "PIN", 1000],
			["r", "Read pin %n", "digital_read", "PIN"],
			[" ", "Log %s to console", "console_log", "Hello, World!"],
			['h', 'When IR sensor detects something', 'ir_sensor_activated', "PIN"],
		],
		"menus": {
			"high_low": ["0", "1"],
			"motor_direction": ["cw", "ccw"],
			"motor_speed" : ["slow", "medium", "fast", "super fast"],
			"train_switch_direction" : ["straight", "turn"]
		},
		url: 'http://www.nowhereville.org'
	};
    
	// Register the extension
	ScratchExtensions.register('felix', descriptor, ext);
})({});

File: felix.py
# no. #!/usr/bin/env python3
# Official docs: https://github.com/llk/scratchx/wiki#command-blocks-that-wait
import json
import os
import sys
import time
import traceback
from subprocess import call
import requests

import pigpio
import RPi.GPIO as GPIO
import psutil
from SimpleWebSocketServer import SimpleWebSocketServer, WebSocket

# It receives messages from the Scratch and reports back any digital input changes.
class FelixServer(WebSocket):
    GPIO.setmode(GPIO.BCM)

    MotorStepCount = 8
    MotorSeq = range(0, MotorStepCount)
    MotorSeq[0] = [0,1,0,0]
    MotorSeq[1] = [0,1,0,1]
    MotorSeq[2] = [0,0,0,1]
    MotorSeq[3] = [1,0,0,1]
    MotorSeq[4] = [1,0,0,0]
    MotorSeq[5] = [1,0,1,0]
    MotorSeq[6] = [0,0,1,0]
    MotorSeq[7] = [0,1,1,0]
    MotorDelayForSpeed = {
        "slow": 40.0/1000,
        "medium": 20.0/1000,
        "fast": 10.0/1000,
        "super fast": 5.0/1000
    }
    
    def handleSetupIRSensor(self, payload):
        self.ir_sensor_pin = int(payload['pin'])
        self.ir_sensor_configured = True;
        self.handleInput(payload) # Also register as an input pin

    def handleSetupMotor(self, payload):
        print("--handleSetupMotor")
        self.motor_pin1 = int(payload['pin1'])
        self.motor_pin2 = int(payload['pin2'])
        self.motor_pin3 = int(payload['pin3'])
        self.motor_pin4 = int(payload['pin4'])
        self.pi.set_mode(self.motor_pin1, pigpio.OUTPUT)
        self.pi.set_mode(self.motor_pin2, pigpio.OUTPUT)
        self.pi.set_mode(self.motor_pin3, pigpio.OUTPUT)
        self.pi.set_mode(self.motor_pin4, pigpio.OUTPUT)
        self.motorConfigured = True;
        self.motorMoving = False;

    def handleRotateMotor(self, payload): # "command": 'rotate_motor', 'speed': speed, 'dir': dir, 'steps': steps
        print("--handleRotateMotor:", payload)
        if not (self.motorConfigured):
            print("Motor not configured; ignoring.")
            return
        
        if (self.motorMoving):
            print("Motor already moving; skipping.")
            return

        self.motorMoving = True
        delay = self.MotorDelayForSpeed.get(payload['speed'], 10)
        if (payload['dir'] == 'cw'):
            self._motorForward(delay, payload['steps'])
        else:
            self._motorBackward(delay, payload['steps'])
        self._setMotorPins(0, 0, 0, 0)
        self.motorMoving = False
    
    def _setMotorPins(self, pin1Val, pin2Val, pin3Val, pin4Val):
        # print("setPins", pin1Val, pin2Val, pin3Val, pin4Val)
        self.pi.write(self.motor_pin1, pin1Val)
        self.pi.write(self.motor_pin2, pin2Val)
        self.pi.write(self.motor_pin3, pin3Val)
        self.pi.write(self.motor_pin4, pin4Val)

    def _motorForward(self, delay, steps):
        # print("--- motorForward", self, delay, steps)
        for i in range(steps):
            for j in range(self.MotorStepCount):
                self._setMotorPins(self.MotorSeq[j][0], self.MotorSeq[j][1], self.MotorSeq[j][2], self.MotorSeq[j][3])
                time.sleep(delay)
                
    def _motorBackward(self, delay, steps):
        # print("--- motorBackward", self, delay, steps)
        for i in range(steps):
            for j in reversed(range(self.MotorStepCount)):
                self._setMotorPins(self.MotorSeq[j][0], self.MotorSeq[j][1], self.MotorSeq[j][2], self.MotorSeq[j][3])
                time.sleep(delay)

    def handleInput(self, payload):
        pin = int(payload['pin'])
        self.pi.set_glitch_filter(pin, 20000)
        self.pi.set_mode(pin, pigpio.INPUT)
        self.pi.callback(pin, pigpio.EITHER_EDGE, self.input_callback)
        
    def handleDigitalWrite(self, payload):
        pin = int(payload['pin'])
        self.pi.set_mode(pin, pigpio.OUTPUT)
        state = payload['state']
        if state == '0':
            self.pi.write(pin, 0)
        else:
            self.pi.write(pin, 1)

    def handleAnalogWrite(self, payload):
        pin = int(payload['pin'])
        self.pi.set_mode(pin, pigpio.OUTPUT)
        value = int(payload['value'])
        self.pi.set_PWM_dutycycle(pin, value)

    def handleServo(self, payload):
        # HackEduca ---> When a user wishes to set a servo:
        # Using SG90 servo:
        # 180 = 2500 Pulses; 0 = 690 Pulses
        # Want Servo 0-->180 instead of 180-->0:
        # Invert pulse_max to pulse_min
        # pulse_width = int((((pulse_max - pulse_min)/(degree_max - degree_min)) * value) + pulse_min)
        # Where:
        # Test the following python code to know your Pulse Range: Replace it in the formula
        # >>>>----------------------->
        # import RPi.GPIO as GPIO
        # import pigpio
        # Pulse = 690 # 0
        # Pulse = 2500 # 180
        # pi = pigpio.pi()
        # pi.set_mode(23, pigpio.OUTPUT)
        # pi.set_servo_pulse_width(23, Pulse)
        # pi.stop()
        # <------------------------<<<<<
        pin = int(payload['pin'])
        self.pi.set_mode(pin, pigpio.OUTPUT)
        value = int(payload['value'])
        DegreeMin = 0
        DegreeMax = 180
        PulseMin = 2500
        PulseMax = 690
        Pulsewidth = int((((PulseMax - PulseMin) / (DegreeMax - DegreeMin)) * value) + PulseMin)
        self.pi.set_servo_pulsewidth(pin, Pulsewidth)
        time.sleep(0.01)

    def handleTone(self, payload):
        pin = int(payload['pin'])
        self.pi.set_mode(pin, pigpio.OUTPUT)

        frequency = int(payload['frequency'])
        frequency = int((1000 / frequency) * 1000)
        tone = [pigpio.pulse(1 << pin, 0, frequency),
                pigpio.pulse(0, 1 << pin, frequency)]

        self.pi.wave_add_generic(tone)
        wid = self.pi.wave_create()

        if wid >= 0:
            self.pi.wave_send_repeat(wid)
            time.sleep(1)
            self.pi.wave_tx_stop()
            self.pi.wave_delete(wid)

    def handleLEDBrightness(self, payload):
        pin = int(payload['pin'])
        brightness = int(payload['brightness']);
        self.pi.set_PWM_dutycycle(pin, brightness)

    def handleLanternBrightness(self, payload):
        brightness = int(payload['brightness'])
        url = "http://10.0.1.5/api/brightness/" + str(brightness)
        print("POSTing: ", url)
        r = requests.post(url = url)
        print("Response: ", r)

    def handleLanternCycleBegin(self, payload):
        url = "http://10.0.1.5/api/mode/cycle"
        print("POSTing: ", url)
        r = requests.post(url = url)
        print("Response: ", r)

    def handleLanternCyclePause(self, payload):
        url = "http://10.0.1.5/api/mode/cycle/pause"
        print("POSTing: ", url)
        r = requests.post(url = url) 
        print("Response: ", r)

    def handleConfigureTrainSwitch(self, payload):
        servoPin = int(payload['switchPin'])
        GPIO.setup(servoPin, GPIO.OUT)
        p = GPIO.PWM(servoPin, 50)
        p.start(0)

    def handleSetTrainSwitch(self, payload):
        print("--handleSetTrainSwitch:", payload)
        servoPin = int(payload['pin'])
        direction = payload['direction']
        duty_cycle = 7.0 if direction == 'straight' else 3.0

        GPIO.setup(servoPin, GPIO.OUT)
        p = GPIO.PWM(servoPin, 50)
        p.start(0)
        p.ChangeDutyCycle(duty_cycle)
        time.sleep(0.5)
        p.ChangeDutyCycle(0)

    def handleConsoleLog(self, payload):
        print("--handleConsoleLog:", payload)
        message = payload['message']
        print("***** **** ***")
        print(message)
        print("***** **** ***")

    def handleMessage(self):
        try:
            payload = json.loads(self.data)
            print("Message received:" + str(payload))
            client_cmd = payload['command']
            print("Client command: " + str(client_cmd))

            if client_cmd == 'input':
                self.handleInput(payload)
            elif client_cmd == 'digital_write':
                self.handleDigitalWrite(payload) 
            elif client_cmd == 'analog_write':
                self.handleAnalogWrite(payload)
            elif client_cmd == 'servo':
                self.handleServo(payload)
            elif client_cmd == 'setup_motor':
                self.handleSetupMotor(payload)
            elif client_cmd == 'rotate_motor':
                self.handleRotateMotor(payload)
            elif client_cmd == 'tone':
                self.handleTone(payload)
            elif client_cmd == 'ready':
                pass
            elif client_cmd == 'lantern_cycle_begin':
                self.handleLanternCycleBegin(payload)
            elif client_cmd == 'lantern_cycle_pause':
                self.handleLanternCyclePause(payload)
            elif client_cmd == 'lantern_brightness':
                self.handleLanternBrightness(payload)
            elif client_cmd == 'led_brightness':
                self.handleLEDBrightness(payload)
            elif client_cmd == 'configure_train_switch':
                self.handleConfigureTrainSwitch(payload)
            elif client_cmd == 'set_train_switch':
                self.handleSetTrainSwitch(payload)
            elif client_cmd == 'console_log':
                self.handleConsoleLog(payload)
            elif client_cmd == 'configure_ir_sensor':
                self.handleSetupIRSensor(payload)
            else:
                print("Unknown command received", client_cmd)
            print("------ - ------")
            print
        except Exception, err:
            print Exception, err
            traceback.print_exc()

    # call back from pigpio when a digital input value changed
    # send info back up to scratch
    def input_callback(self, pin, level, tick):
        payload = {'report': 'digital_input_change', 'pin': str(pin), 'level': str(level)}
        print('callback', payload)
        msg = json.dumps(payload)
        self.sendMessage(msg)

        # Also send ir-sensor-triggered if this is a registered IR sensor
        if (self.ir_sensor_configured and (pin == self.ir_sensor_pin)):
            payload = {'report': 'ir_sensor_triggered', 'state': str(level)}
            print('callback', payload)
            msg = json.dumps(payload)
            self.sendMessage(msg)

    def handleConnected(self):
        self.pi = pigpio.pi()
        print(self.address, 'connected')

    def handleClose(self):
        print(self.address, 'closed')

def run_server():
    # checking running processes.
    # if the backplane is already running, just note that and move on.
    found_pigpio = False
    found_scratch = False
    print('Starting server')

    for pid in psutil.pids():
        p = psutil.Process(pid)
        if p.name() == "pigpiod":
            found_pigpio = True
            print("pigpiod is running")
        else:
            continue

    if not found_pigpio:
        call(['sudo', 'pigpiod'])
        print('pigpiod has been started')
        
    for pid in psutil.pids():
        p = psutil.Process(pid)
        if p.name() == "scratch2":
            found_scratch = True
            print("scratch2 is running")
        else:
            continue

    if not found_scratch:
        os.system('scratch2&')
        print('scratch2 has been started')

    server = SimpleWebSocketServer('', 9001, FelixServer)
    server.serveforever()

if __name__ == "__main__":
    try:
        run_server()
    except KeyboardInterrupt:
        sys.exit(0)
